### PR TITLE
Update dependencies and enable new rules to support TypeScript 4 and ESLint 7

### PR DESCRIPTION
**Pull request by iansu**
- Update dependencies to support TypeScript 4, ESLint 7 and Prettier 2
- Add `config-backend-next` and `config-frontend-next` with rules being considered for a future release
- Enable `@typescript-eslint/no-unused-before-define` rule as error in TypeScript files
- Enable `@typescript-eslint/prefer-optional-chain` rule as warning
- Enable `unicorn/no-reduce` rule as warning
- Enable `unicorn/prefer-optional-catch-binding` rule as warning
- Disable `@typescript-eslint/explicit-module-boundary-types` rule in test files
- Disable `@typescript-eslint/no-empty-function` rule in test files

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,17 @@
 # Changelog
 
+## 0.6.0 (November 20, 2020)
+
+- Update dependencies to support TypeScript 4, ESLint 7 and Prettier 2
+- Enable `@typescript-eslint/no-unused-before-define` rule in TypeScript files
+- Enable `@typescript-eslint/prefer-optional-chain` rule
+- Enable `unicorn/no-null` rule
```
Comment (usernameseb) [inline CHANGELOG.md:8]: Does this one still allow null to be used for return types?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,17 @@
...
+- Enable `unicorn/no-reduce` rule
```
Comment (usernameseb) [inline CHANGELOG.md:9]: I do believe we use reduce in areas of statement generation, but I agree it's not readable.

**Review by usernameseb — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (iansu) [reply CHANGELOG.md:8]: Why are you returning `null`? I am a little on the fence about this one. In general we should prefer `undefined` over `null` but sometimes you do have to use `null`. It's only a warning but I have already disabled it in a couple files.

File: CHANGELOG.md (hunk 2 above)
Comment (iansu) [reply CHANGELOG.md:9]: This is just a warning. Sometimes it does make sense to use `reduce` and you can just disable the rule.

**Review by ronniemagatti — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (usernameseb) [reply CHANGELOG.md:8]: Yeah maybe we can just disable it in the offending places where we can't use `undefined`, just thinking about providers that are still returning `document | null`, we've also got graphql `__resolveType` functions that always return `string | null`, etc.

Comment (iansu) [reply CHANGELOG.md:9]: I just added `next` versions of the configs so maybe I'll move this rule in there so we can preview it before adopting it.

**Pull request merged**